app/models/sample_manifest/plate_behaviour.rb

```diff
@@ -83,65 +74,60 @@ def labware=(labware)
-      cloned_well_data = well_data.dup
-      plates.each { |plate| yield(cloned_well_data.slice!(0, plate.size), plate) }
-    end
-
-    def labware_from_barcodes
-      Labware.with_barcode(barcodes)
+    def generate_plates(purpose)
+      Array.new(count) { purpose.create!(:without_wells) }.sort_by(&:human_barcode)
     end
 
-    def generate_wells_asynchronously(map_ids_to_sample_ids, plate_id)
-      Delayed::Job.enqueue SampleManifest::GenerateWellsJob.new(@manifest.id, map_ids_to_sample_ids, plate_id)
+    def insert_sanger_sample_ids
+      sanger_sample_ids = generate_sanger_ids(@plates.sum(&:size) * @manifest.rows_per_well)
+      sanger_sample_ids.map do |sanger_sample_id|
+        SangerSampleId.generate_sanger_sample_id!(study.abbreviation, sanger_sample_id)
+      end
     end
 
-    # rubocop:todo Metrics/MethodLength
-    def generate_plates(purpose) # rubocop:todo Metrics/AbcSize
-      study_abbreviation = study.abbreviation
-
-      well_data = []
-      plates = Array.new(count) { purpose.create!(:without_wells) }.sort_by(&:human_barcode)
-
-      plates.each do |plate|
-        sanger_sample_ids = generate_sanger_ids(plate.size)
+    # output:
+    # plate_id => { map_id => [sanger_sample_id, sanger_sample_id, ...] }
+    def build_well_data(sanger_sample_ids)
+      @plates.each_with_object({}) do |plate, well_data|
+        well_data[plate.id] = {}
 
         plate.maps.in_column_major_order.each do |well_map|
-          sanger_sample_id = sanger_sample_ids.shift
-          generated_sanger_sample_id = SangerSampleId.generate_sanger_sample_id!(study_abbreviation, sanger_sample_id)
-
-          well_data << [well_map, generated_sanger_sample_id]
+          well_data[plate.id][well_map.id] = sanger_sample_ids.shift(@manifest.rows_per_well)
         end
       end
+    end
 
-      generate_wells(well_data, plates)
-      @manifest.update!(barcodes: plates.map(&:human_barcode))
+    # Each of the plates is handled by an individual job.
+    # If it doesn't do this we run the risk that the 'handler' column in the database
+    # for the delayed job will not be large enough and will truncate the data.
+    def build_wells_async(well_data)
+      @plates.each do |plate|
+        Delayed::Job.enqueue SampleManifest::GenerateWellsJob.new(@manifest.id, well_data[plate.id], plate.id)
+      end
+    end
 
-      plates
+    # output:
+    # [{barcode, position, sanger_sample_id}, {barcode, position, sanger_sample_id}, ...]
+    def build_details_array(well_data)
+      @details_array =
+        @plates.flat_map do |plate|
+          well_data[plate.id].flat_map do |map_id, sanger_sample_ids|
+            sanger_sample_ids.map do |sanger_sample_id|
+              {
+                barcode: plate.human_barcode,
+                position: plate.maps.find(map_id).description,
+                sample_id: sanger_sample_id
+              }
+            end
+          end
+        end
```

Comment:
It looks like perhaps the indenting has lost its way here.

Reply:
Hmm, Prettier seems to want it like this. It's because there's a new line after the `=`:

```
@details_array =
        @plates.flat_map do |plate|
```

So the `end` lines up with `@plates`

Reply:
Ah right, I see.  I just saw the `end` statements not being one tab-width apart and thought it was off.  It niggles me when this happens, but you can't fight the code styles unless we change the rules it uses, which is more of a headache usually.